coroutines-guide.md

```diff
@@ -2022,79 +2025,164 @@ The locking in this example is fine-grained, so it pays the price. However, it i
 
 <!--- TEST ARBITRARY_TIME
+Preparing to send a lot of inc requests
+CountingActor started
 Completed 1000000 actions in xxx ms
 Counter = 1000000
 -->
 
-It does not matter (for correctness) what context the actor itself is executed in. An actor is
+Note that the same actor, but without `onStart` handler, can be implemented using more compact [actor] builder: 
+```kolin
+fun createCountingActor(): TypedActor<CounterMsg> = actor { message ->
+    when (message) {
+        is IncCounter -> counter++
+        is GetCounter -> message.response.complete(counter)
+    }
+}
+```
+
+ `counter` is not protected by any synchronization, because actor processes messages sequentially one by one.
+It does not matter for correctness what context the actor itself is executed in. An actor is
 a coroutine and a coroutine is executed sequentially, so confinement of the state to the specific coroutine
 works as a solution to the problem of shared mutable state. Indeed, actors may modify their own private state, but can only affect each other through messages (avoiding the need for any locks).
 
 Actor is more efficient than locking under load, because in this case it always has work to do and it does not 
 have to switch to a different context at all.
 
-> Note, that an [actor] coroutine builder is a dual of [produce] coroutine builder. An actor is associated 
-  with the channel that it receives messages from, while a producer is associated with the channel that it 
-  sends elements to.
+### Typesafe actors
+What is usually counterintuitive about Actor-based API is the absence of meaningful method names and requirement to introduce `sealed` classes for every actor
+which processes more than one type of messages. And in any real-world project it's very likely that any actor can process more than only one kind of requests.
+For this purposes we have [Actor] base class. With [Actor], any method with any arguments can act like `send` method of `TypedActor`, the only requirement is to define methods
+using [act][Actor.act].
+
+Let's implement the same `CountingActor` as in the previous example;  
+
+<!--- INCLUDE
+import kotlinx.coroutines.experimental.actors.*
+import kotlinx.coroutines.experimental.guide.sync01.*
+-->
+
+```kotlin
+class CountingActor : Actor() {
+    
+    private var counter: Int = 0
+    
+    override suspend fun onStart() {
+      println("CountingActor started")
+    }
+    
+    suspend fun increment() = act { // <- note act {} extension
+      counter++
+    }
+    
+    suspend fun counter(response: CompletableDeferred<Int>) = act {
+          response.complete(counter)
+    }
+}
+
+fun main(args: Array<String>) = runBlocking<Unit> {
+    val counter = CountingActor() // create the actor
+    println("Preparing to send a lot of inc requests")
+    massiveRun(CommonPool) {
+        counter.increment()
+    }
+    
+    val response = CompletableDeferred<Int>()
+    counter.counter(response)
+    println("Counter = ${response.await()}")
```

Comment:
It's a matter of taste.
For me this pattern is obscure

Reply:
Agree. We should actually rewrite actor example in the guide. Actors should never be written like that. You should never need to wait for a response to a particular actor's action.

Reply:
Same example here

https://github.com/Kotlin/kotlinx.coroutines/blob/master/coroutines-guide.md#actors

Following your reasoning the "counter" use case fits really better using a `Mutex`, otherwise you have to consider the feature of replying actor's messages.